.github/workflows/ci.yml

```diff
@@ -106,21 +117,90 @@ jobs:
     - uses: actions/setup-node@v1
       with:
         node-version: ${{ secrets.NODE_VERSION }}
-    - uses: actions/cache@v1
+    - name: Cache node modules
+      id: node-cache
+      uses: actions/cache@v2
       with:
-        path: ~/.npm
-        key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
-        restore-keys: |
-          ${{ runner.os }}-node-
-    - run: |
+        path: |
+          ~/cache
+          ~/.cache
+          !~/cache/exclude
+          **/node_modules
+        key: ${{ runner.os }}-node-${{ hashFiles('**/lockfiles') }}
+    - name: Install dependencies (if needed)
+      if: steps.node-cache.outputs.cache-hit != 'true'
+      run: |
         cd ui
-        npm i
-        npm test
+        npm ci
+        cd ..
+        cd provider-ui
+        npm ci
+    - name: Run tests
+      run: |
+        cd ui
+        npm run ci-test-integration
         cd ..
         cd provider-ui
-        npm i
         npm test
         cd ..
+  tests-ui-e2e:
+    needs: [build-backend, build-ui]
+    name: UI end-to-end tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@master
+        with:
+          fetch-depth: 1
+      - name: Download artifacts
+        uses: actions/download-artifact@v2
+        with:
+          name: meshery
+      - name: Setup go
+        uses: actions/setup-go@v1
+        with:
+          go-version: ${{ secrets.GO_VERSION }}
+      - name: Setup node
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ secrets.NODE_VERSION }}
+      - name: Cache node modules
+        id: node-cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/cache
+            ~/.cache
+            !~/cache/exclude
+            **/node_modules
+          key: ${{ runner.os }}-node-${{ hashFiles('**/lockfiles') }}
+          restore-keys: |
+            ${{ runner.os }}-node-
+      - name: Install dependencies (if needed)
+        if: steps.node-cache.outputs.cache-hit != 'true'
+        run: |
+          cd ui
+          npm ci
+          cd ..
+          cd provider-ui
+          npm ci
+      - name: Build Next App
+        run: cd ui && npm run build && npm run export
+      - name: Run backend
+        env:
+          SAAS_BASE_URL: http://localhost:9876
+          PORT: 9081
+          ADAPTER_URLS: "mesherylocal.layer5.io:10000 mesherylocal.layer5.io:10001 mesherylocal.layer5.io:10002 mesherylocal.layer5.io:10003 mesherylocal.layer5.io:10004 mesherylocal.layer5.io:10008 mesherylocal.layer5.io:10009"
```

Comment:
I wonder if there's an alternative way of loading these from the `Makefile` or `docker-compose.yaml` such that this list doesn't need to be maintained (new adapters are not added frequently, but are added occasionally).

If not, then, we may want to expand this list to include ports 10007 (kuma) - it's ready for testing. Almost ready to move from beta to stable.

Reply:
Ok, I see. Let me see if we could extract them from those files. If not, I'll add 10007 as you mention.

Reply:
You can just use `make run-local` over here, which will save us from adding artifacts action.

Reply:
I didn't use `make run-local` because that command [builds the _meshery_ binary too](https://github.com/layer5io/meshery/blob/master/Makefile#L63); and we already building _meshery_ in a previous step ([build-backend](https://github.com/layer5io/meshery/blob/master/.github/workflows/ci.yml#L72)). Using `artifacts` - _a way to share files between jobs_ -  will allow us to not build _meshery_ twice. If we rebuild the _meshery_ again, it will take **50s** instead of **1s** downloading the artifact. You can see that in [this Github Action](https://github.com/layer5io/meshery/pull/1862/checks?check_run_id=1361845268). This is like using caching node dependencies to avoid install them twice.

Reply:
I added the port directly _(10007)_ because actually, we don't have tests related to test those Adapters yet. I think that if there is a new Adapter, it'll be included as part of writting the test. What do you' all think?